MIPs/mip-3.md

```diff
@@ -0,0 +1,184 @@
+
+  ```markdown
+  if (from.length > 1) {
+  throw rpcErrors.invalidParams(
+  'Currently we do not support multiple tokens swap',
+  );
+  }
+  ```
+
+- **Inactive or Unsupported Swaps**: If the swap is inactive or is not possible on the current chain, an alert with the message 'Swap is not active or is not possible on this chain' will be triggered. See more details about error codes in the api-spec: https://github.com/MetaMask/api-specs/pull/201
+
+  ```markdown
+  const isSwappable = isSwapsAllowed(chainId) && swapsIsLive;
+  if (!isSwappable) {
+  Alert.alert(`Swap is not available on this chain ${networkName}`);
+  throw rpcErrors.methodNotSupported(
+  `Swap is not available on this chain ${networkName}`,
+  );
+  }
+  ```
+
+5. Security: Dapp Developers should ensure that the token addresses provided in the method call are secure and valid for the current chain ID. They should also inform users about the potential risks and considerations when performing token swaps.
+   On MetaMask platforms we have warnings in place that say to the user to be carefull when a token is not that trustworthy. We check if the number of occurrences is more than one, If not we show an alert the we obligate the user to ready and press a button to be able to swap.
+
+6. Testing: Before deploying the `wallet_swapAsset` method in a live environment, MetaMask engineers should thoroughly test it to ensure it works correctly and handles errors appropriately. This includes testing with different token types, amounts, and network conditions.
+
+7. Interoperability: The `wallet_swapAsset` method provides a standardized way for dApps to request token swaps from wallets. This can improve interoperability between different dApps and wallets, as they can all use the same method for token swaps.
+
+## User Experience Considerations
+
+Users will now be able to swap tokens without leaving dapps.
+Since `wallet_swapAsset` method will be new to some users, providing educational resources may help in understand how the method works as well as how to use it.
+If the `wallet_swapAsset` method encounters an error, the wallet/dapp should provide a clear and understandable error message to the user. This helps users understand what went wrong and how to fix it.
+
+While using MetaMask Mobile with this method, the mobile app will first check if selected account is different between the wallet and the dApp. If it is different, the mobile app will switch to the matching account before continuing with the swap.
+
+```markdown
+if (
+safeToChecksumAddress(selectedAddress) !==
+checksummedDappConnectedAccount
+) {
+Engine.context.PreferencesController.setSelectedAddress(
+checksummedDappConnectedAccount,
+);
+}
+```
```

Comment:
Is there a reason we are saying we'll only do this in mobile? I guess is it not required in extension?

Reply:
On mobile it's needed because you can have a selected account on the wallet view and a different selected account on the dapp on in-app browser, and what json rpc does now is redirect the user to the swaps flow, so if the address between the wallet view and in app browser dapp is different, we change the account on the wallet view to the one at the dapp.


On extension this doesn't happen.

Reply:
Added here: https://github.com/tommasini/metamask-improvement-proposals/blob/e3f72449345cddb460f76c60df53a140b1560c0d/MIPs/mip-3.md#L157